czkawka_core/src/duplicate.rs

```diff
@@ -344,112 +340,156 @@ impl DuplicateFinder {
-            if stop_receiver.is_some() && stop_receiver.unwrap().try_recv().is_ok() {
-                // End thread which send info to gui
-                progress_thread_run.store(false, Ordering::Relaxed);
-                progress_thread_handle.join().unwrap();
-                return false;
-            }
+        let sst: SystemTime = SystemTime::now();
 
-            let current_folder = folders_to_check.pop().unwrap();
+        let allowed_extensions = Arc::new(self.allowed_extensions.file_extensions.clone());
+        let minimal_file_size = self.minimal_file_size.clone();
+        let maximal_file_size = self.maximal_file_size.clone();
 
-            // Read current dir, if permission are denied just go to next
-            let read_dir = match fs::read_dir(&current_folder) {
-                Ok(t) => t,
-                Err(_) => {
-                    self.text_messages.warnings.push(format!("Cannot open dir {}", current_folder.display()));
+        let (tx, rx): (std::sync::mpsc::Sender<PathBuf>, std::sync::mpsc::Receiver<PathBuf>) = std::sync::mpsc::channel();
+        let collect_thread = std::thread::spawn(move || {
+            let mut vec: Vec<(String, FileEntry)> = Vec::new();
+            for path_buf in rx {
+                if path_buf.is_dir() {
                     continue;
-                } // Permissions denied
-            };
-
-            // Check every sub folder/file/link etc.
-            'dir: for entry in read_dir {
-                let entry_data = match entry {
-                    Ok(t) => t,
-                    Err(_) => {
-                        self.text_messages.warnings.push(format!("Cannot read entry in dir {}", current_folder.display()));
-                        continue 'dir;
-                    } //Permissions denied
-                };
-                let metadata: Metadata = match entry_data.metadata() {
-                    Ok(t) => t,
-                    Err(_) => {
-                        self.text_messages.warnings.push(format!("Cannot read metadata in dir {}", current_folder.display()));
-                        continue 'dir;
-                    } //Permissions denied
+                }
+                let file_name_lowercase = match path_buf.file_name() {
+                    Some(t) => t.to_string_lossy().to_lowercase(),
+                    None => continue,
                 };
-                if metadata.is_dir() {
-                    if !self.recursive_search {
-                        continue 'dir;
-                    }
 
-                    let next_folder = current_folder.join(entry_data.file_name());
-                    if self.directories.is_excluded(&next_folder) {
-                        continue 'dir;
+                // Checking allowed extensions
+                if !allowed_extensions.is_empty() {
+                    let allowed = allowed_extensions.iter().any(|e| file_name_lowercase.ends_with((".".to_string() + e.to_lowercase().as_str()).as_str()));
```

Comment:
I believe this line has too much complexity. You may lift it by moving code to the `Extensions` abstraction like this:

```rust
impl Extensions {
    fn contains(&self, file_name_lowercase: &str) -> bool {
        self.file_extensions.iter().any(|e| {
            let dot_then_ext = ".".to_string() + e.to_lowercase().as_str();
            file_name_lowercase.ends_with(dot_then_ext.as_str())
        });
    }
}

// ...
if !self.allowed_extensions.contains(&file_name_lowercase[..]) {
    // Not an allowed extension, ignore it.
    continue;
}
```

Reply:
my bad, use a more descriptive function name like `matches_filename`, not `contains`

Reply:
This could be optimized, by adding when necessary dots at the start when necessary and use `to_lowercase` to `set_allowed_extensions` function.

But this should be done in new PR, because I'm not sure when I merge this PR, because in 2 implementations of threads which I wrote, I got smaller performance than without.

Reply:
I could volunteer to do such a simple task.

Reply:
I feel like this is premature optimization -- on the other hand, this code is O(m*n) in the number of files (m) times the number of extensions (n) so we could use some kind of associative set data structure here.

Reply:
So this is a low hanging fruit up for optimization.